File: day04/star08.py
from collections import Counter

def parse_line(line):
    winningstr, havestr = line.split(':')[1].split('|')
    winning = set(winningstr.split())
    have = set(havestr.split())
    both = have.intersection(winning)
    return len(both)

with open('input.txt') as f:
    cards = f.readlines()

c = Counter(range(len(cards)))

for i in c.keys():
    for j in range(parse_line(cards[i])):
        c[i+j+1] += c[i]

print(sum(c.values()))

File: day04/star08.rs
use std::{fs, io::BufRead, collections::{HashSet, HashMap}};

fn open_file(path: &str, dir: &str) -> std::fs::File {
    return match fs::File::open(path) {
        Ok(f) => f,
        Err(_) => fs::File::open(dir.to_string() + path).expect("File not found!"),
    };
}

fn parse_line(line: String) -> usize {
    let (winningstr, havestr) = line.split_once(':').unwrap().1
                                                .split_once('|').unwrap();
    let winning: HashSet<_> = winningstr.split_whitespace().collect();
    let have: HashSet<_> = havestr.split_whitespace().collect();

    let num: usize = winning.intersection(&have).collect::<HashSet<_>>().len();
    return num;
}

fn main() {
    let f = open_file("input.txt", "day04/");
    let reader = std::io::BufReader::new(f);
    let cards: Vec<usize> = reader.lines()
                   .map(|line| parse_line(line.unwrap()))
                   .collect();

    let mut c: HashMap<_, u32> = (0..cards.len()).map(|key| (key, 1)).collect();
    for i in 0..(c.len()) {
        let ci = *c.get(&i).unwrap();
        for j in 0..(cards[i]) {
            *c.entry(i+j+1).or_insert(1) += ci;
        }
    }
    println!("{}", c.into_values().sum::<u32>());
}